### DISCUSSION TITLE
How do I remove the manifest from iOS to allow OAuth redirect to work?

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#20119] [Help]
I am using [next-auth](https://github.com/vercel/next.js/tree/canary/examples/with-next-auth) with the [progressive-web-app](https://github.com/vercel/next.js/tree/canary/examples/progressive-web-app) example, but authentication doesn't work on iOS. 

I created a [repo](https://github.com/Ramos-Diego/pwa) which is the simplest combination of these two examples. It's available at [pwa-sage.vercel.app](https://pwa-sage.vercel.app/).

I found a solution in another [repo](https://github.com/GoogleChromeLabs/pwacompat/issues/15), but I am unable to successfully implement it.

The problem is described in [this article](https://medium.com/@firt/whats-new-on-ios-12-2-for-progressive-web-apps-75c348f8e945). What I understood is that Safari doesn't handle redirects like other browsers thus breaking OAuth.

What exact changes do I need to make in `manifest.json` or application in order to use next-auth with progressive-web-app and still have OAuth working in iOS? I don't mind removing all PWA support in iOS devices.

**Comment by Tomburgs** [comment]
I tested your app as is and I didn't have any issues getting redirected to Google's Sign in screen (Tested both in Simulator and on actual iPhone).

The issue you linked also says that the issue was fixed in iOS 12.2.0, could perhaps you be using an older iOS version?

If so, to implement the fix defined in said issue, you would do as follows in `pages/_app.js`:
```js
export default function MyApp({ Component, pageProps }) {
  ...
  const isIOS = typeof navigator !== 'undefined' && /iPhone/.test(navigator.platform);

  return (
    <>
      <Head>
        <link rel={ isIOS ? 'no-on-ios' : 'manifest' } href="/manifest.json" />
        ...
      </Head>
      ...
    </>
  );
}
```

**Reply by ghost** [reply]
I did not explain the problem correctly. I can get to that page too. The problem is that I can't go back to my app from Google's Authentication prompt. I implemented your solution, but it did not work. 

If you do not trust entering you Google credentials in my test application, you can deploy it yourself and test it all the way through.

You will need these Environment Variables:
```
NEXTAUTH_URL
NEXTAUTH_GOOGLE_ID
NEXTAUTH_GOOGLE_SECRET
NEXTAUTH_JWT_SIGNING_PRIVATE_KEY (Optional)
NEXTAUTH_JWT_SECRET (Optional)
NEXTAUTH_SECRET (Optional)
```

You will get this error if you check the console for the `/api/auth/[...nextauth]` function in Vercel:

```bash
[GET] /api/auth/callback/google?state=a18d41d8...bd&code=4%2...lHx1JC-fg&scope=email+profile+openid+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email&authuser=0&prompt=none
12:57:40:47
Status: 302
Duration: 3.58ms
Memory Used: 108 MB
ID: l...d-160...1-79c...4
User Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 14_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Mobile/15E148 Safari/604.1 2020-12-12T17:57:40.487Z	a...c-690c-4ac7-b...7-f8...82	ERROR	[next-auth][error][callback_oauth_error] Error: Invalid state returned from oAuth provider
    at /var/task/node_modules/next-auth/dist/server/lib/oauth/callback.js:46:27
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (/var/task/node_modules/next-auth/dist/server/lib/oauth/callback.js:26:103)
    at _next (/var/task/node_modules/next-auth/dist/server/lib/oauth/callback.js:28:194)
    at /var/task/node_modules/next-auth/dist/server/lib/oauth/callback.js:28:364
    at new Promise (<anonymous>)
    at /var/task/node_modules/next-auth/dist/server/lib/oauth/callback.js:28:97
    at /var/task/node_modules/next-auth/dist/server/lib/oauth/callback.js:143:17
    at /var/task/node_modules/next-auth/dist/server/routes/callback.js:58:31
    at Generator.next (<anonymous>) 
https://next-auth.js.org/errors#callback_oauth_error
```

As you can see I was using `iPhone OS 14_2`, which is the latest version of iOS. I get the same error when attempting to login from Safari 13 in a `Mac OS X 10_15`

```
User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Safari/605.1.15
```

**Comment by ghost** [comment]
The only solution that I have found thus far is disabling the OAuth state support for the provider: 

```js
// [...nextauth].js
Providers.Google({
  clientId: process.env.NEXTAUTH_GOOGLE_ID,
  clientSecret: process.env.NEXTAUTH_GOOGLE_SECRET,
  state: false, // Disable the state feature
}),
```
I am not qualified to say if this a safe solution. Apparently, this opens a [CSRF attacks](https://auth0.com/docs/protocols/state-parameters) vulnerability.
